### DISCUSSION TITLE
Query Protocol: reset comes after initialization, but before use

### DISCUSSION THREAD

**Discussion by sploders101** [apache/couchdb#2216] [Q&A]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
The documentation for the query server protocol suggests that ```["reset", {...}]``` is supposed to reset the server and cause it to **forget all previous input**. Because of this, I have put all mutable state variables in a separate object so that they can be easily reset, including the design document registry. Unfortunately, upon installing this into CouchDB, I was getting errors that design documents were missing from the registry, and upon further investigation, found that CouchDB was inserting the design documents, deleting them, and then trying to run the functions in them.

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

## Steps to Reproduce
Unfortunately, I do not currently have permission to release the code for this. However, by setting
```ini
[log]
level = debug
```
you should be able to see similar results. Some data has been omitted for the sake of brevity.

```
[debug] ... Input  :: ["ddoc","new","_design/validation",{"_id":"_design/validation","_rev":"11-34bc995353612a621f30a4df37900600","validate_doc_update":"function(){...}"}]
[debug] ... Output :: true
[debug] ... Input  :: ["reset",{"reduce_limit":true,"timeout":5000}]
[debug] ... Output :: true
[debug] ... Input  :: ["ddoc","_design/validation",["validate_doc_update"],[{"_id":"S00-00000","_rev":"25-4f0444901f5a3bcfa2a16d27d39dc62f", ...},{"_id":"S00-00000","_rev":"25-4f0444901f5a3bcfa2a16d27d39dc62f", ...},{"db":"production_orders","name":"devtest","roles":["admin"]},{}]]
[debug] ... Output :: 1
```
I have since removed the registry from the resettable state (hence no error in the console), but this does not seem like a good solution to me, as I would have to restart the database to remove a design document.

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

I would expect the reset to come before the document is transferred, or after the function is run. Ex:
```
[debug] ... Input  :: ["reset",{"reduce_limit":true,"timeout":5000}]
[debug] ... Output :: true
[debug] ... Input  :: ["ddoc","new","_design/validation",{"_id":"_design/validation","_rev":"11-34bc995353612a621f30a4df37900600","validate_doc_update":"function(){...}"}]
[debug] ... Output :: true
[debug] ... Input  :: ["ddoc","_design/validation",["validate_doc_update"],[{"_id":"S00-00000","_rev":"25-4f0444901f5a3bcfa2a16d27d39dc62f", ...},{"_id":"S00-00000","_rev":"25-4f0444901f5a3bcfa2a16d27d39dc62f", ...},{"db":"production_orders","name":"devtest","roles":["admin"]},{}]]
[debug] ... Output :: 1
```
--or--
```
[debug] ... Input  :: ["ddoc","new","_design/validation",{"_id":"_design/validation","_rev":"11-34bc995353612a621f30a4df37900600","validate_doc_update":"function(){...}"}]
[debug] ... Output :: true
[debug] ... Input  :: ["ddoc","_design/validation",["validate_doc_update"],[{"_id":"S00-00000","_rev":"25-4f0444901f5a3bcfa2a16d27d39dc62f", ...},{"_id":"S00-00000","_rev":"25-4f0444901f5a3bcfa2a16d27d39dc62f", ...},{"db":"production_orders","name":"devtest","roles":["admin"]},{}]]
[debug] ... Output :: 1
[debug] ... Input  :: ["reset",{"reduce_limit":true,"timeout":5000}]
[debug] ... Output :: true
```
with the latter obviously needing the design document to be re-sent.

[NOTE]: # ( Tell us what you expected to happen. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB Version used: 2.3.1
* Browser name and version: N/A
* Operating System and version: Linux docker - couchdb:2.3.1

## Additional context
It is possible that this is intended (since the default query server works fine), and the documentation needs changed instead, but I believe there should be a way to remove design documents from memory without restarting the server.

[TIP]:  # ( Add any other context about the prbolem here. )

**Comment by wohali** [comment]
Hey @sploders101 , sorry for the late reply here.

Here's what `reset` does in our JavaScript query server: https://github.com/apache/couchdb/blob/master/share/server/state.js#L13-L21

As part of that same change request, whenever a `require` occurs, a brand new sandbox is created and the function is brought into that sandbox: https://github.com/apache/couchdb/commit/33a71410ea7925ffa2da734e73db460d29f68cf7#diff-5426946b592c97e2fc147027d857d90cR67-R100

Does this answer your question?
